quickinput-sample/src/promptCommandWithHistory.ts

```diff
@@ -0,0 +1,104 @@
+import { window, Disposable, Memento } from 'vscode';
+import { QuickPickItem } from 'vscode';
+
+export const historyPath = `${process.env.HOME}/.vscode-cmd-history`;
+
+const HISTORY_KEY = 'HISTORY_KEY';
+
+/**
+ * A command prompt with history
+ * 
+ */
+export const getPromptCommand = (memo: Memento) => {
+	const pickCommand = getPickCommand(memo);
+	// §TODO: cache promptCommand!
+	return async function promptCommand() {
+		const command = await pickCommand();
+		if (command) {
+			window.showInformationMessage(`You picked the following command: '${command}'`)
+		}
+	}
+}
+
+abstract class CommandItem implements QuickPickItem {
+	public label: string;
+	public description?: string;
+	public abstract type: string;
```

Comment:
You're using `type === 'input'` and `item instanceof InputItem` (or `HistoryItem`) which is two approaches to the same problem. Let's only use `type` and instead of having the small class hierarchy, let's have only a single interface `CommandItem`. What do you think?

Reply:
good point :)